cli/azd/pkg/environment/environment.go

```diff
@@ -36,14 +37,78 @@ const ContainerRegistryEndpointEnvVarName = "AZURE_CONTAINER_REGISTRY_ENDPOINT"
 const ResourceGroupEnvVarName = "AZURE_RESOURCE_GROUP"
 
 type Environment struct {
-	// Values is a map of setting names to values.
-	Values map[string]string
+	// values is a map of setting names to values.
+	// all keys are converted to upper case before it is set.
+	values map[string]string
 	// File is a path to the file that backs this environment. If empty, the Environment
 	// will not be persisted when `Save` is called. This allows the zero value to be used
 	// for testing.
 	File string
 }
 
+// DeleteVariable removes an entry from azd environment which key matches the variable name.
+func (e *Environment) DeleteVariable(variableName string) {
+	delete(e.values, strings.ToUpper(variableName))
+}
+
+// SetVariable update or create an entry to the azd environment.
+// key is converted to upper case.
+func (e *Environment) SetVariable(key, value string) {
+	e.values[strings.ToUpper(key)] = value
+}
+
+// GetValue get the value of an entry with the requested key.
+// return empty string and false if the key is not found.
+// Use ValueOf if you don't need to check if value is found
+func (e *Environment) GetValue(key string) (string, bool) {
+	value, found := e.values[strings.ToUpper(key)]
+	return value, found
+}
+
+// GetValue get the value of an entry with the requested key.
+// return empty string and false if the key is not found.
+// Use GetValue to return if the value is found as well
+func (e *Environment) ValueOf(key string) string {
+	return e.values[strings.ToUpper(key)]
+}
+
+// HasValue return true if the key is found and the value is not empty string.
+func (e *Environment) HasValue(key string) bool {
+	value, found := e.values[strings.ToUpper(key)]
+	return found && value != ""
+}
+
+// Init make sure that values has an empty map.
+func (e *Environment) Init() {
+	if e.values == nil {
+		e.values = make(map[string]string)
+	}
+}
+
+// CopyValues get a copy of the azd environment variable entries.
+func (e *Environment) CopyValues() map[string]string {
+	if e.values == nil {
+		return nil
+	}
+
+	result := make(map[string]string)
+	for k, v := range e.values {
+		// keys are already upper case
+		result[k] = v
+	}
+	return result
+}
+
+// ToStringArray create an array of strings where each string is in the form of
+// "key=value"
+func (e *Environment) ToStringArray() []string {
```

Comment:
Is this potentially duplication over `godotenv.Marshal` in `Save`? Otherwise, might need to document the difference

Reply:
very similar, but not the same.
`godotenv.Marshal()` returns a string out of the `map[string]string`.  We use it as part of the `EnvVarsFormatter` 

In case of `ToStringArray()`, we return an array `[]string` out of the `map[string]string`.
This is used by the `package()` method from a `framework_service_npm`.  The method needs to add more values to the output without writing to the environment, like:

```go
        // Run Build, injecting env.
	envs := np.env.ToStringArray()
	envs = append(envs, "NODE_ENV=production")
```